### ISSUE TITLE
Create best objective watch for Cooperative PredPrey

### ISSUE THREAD

**Issue by Rollinsa** [schrum2/MM-NEATv2#145]
#17 #142 Create a best objective watch (postBestObjectiveWatch and postAllBestObjectiveWatch) for Cooperative PredPrey because the original classes do not apply for predPrey coevolution (and pacman's cooperation coevolution does not either)

**Comment by Rollinsa** [comment]
Also make sure that the graphs that are plotted according to the fitness throughout evolution are correct (correctly display each agent/population's individual fitness scores) for PredPrey cooperative coevolution for each population/agent being evolved.

**Comment by schrum2** [comment]
Do this after #146. 

Make a class called ObjectiveBestTeamsExperiment that is modeled on ObjectiveBestNetworksExperiment but is for a cooperative coevolution experiment. Given the files created by the code from #146, you should be able to figure out which files to load.

Note that ObjectiveBestNetworksExperiment is actually used by both postBestObjectiveWatch and postAllBestObjectiveWatch. Lauren actually helped expand ObjectiveBestNetworksExperiment to support postAllBestObjectiveWatch.

Once you complete ObjectiveBestTeamsExperiment, create batch files postTeamBestObjectiveWatch and postTeamAllBestObjectiveWatch to use it.

**Comment by schrum2** [comment]
Actually, first try making a batch file that uses the experiment AllCooperativeCoevolutionTeamsExperiment. This should work already (even before completing #146) but it might show you more than you want to see. It takes the Pareto fronts from each of the final populations and makes every possible team combination out of these genotypes. See how it works!

The other batch files could still be useful though.

**Comment by Rollinsa** [comment]
Created some comments/TODO's/templates for future work to fix the current cooperative coevolution post best watch and to be able to implement a more useful cooperative coevolution one for my torusPredPrey purposes (one with a lot less being shown).

73052eb6b6f8dc1e8deb4d097e6f7d50b753aeca

**Comment by Rollinsa** [comment]
Fixed an issue/error in the cooperativeTorusPredPrey task which was causing errors when running with monitorInputs turned on or when trying to use post best objective watches. I am still unsure if this will be a completely functional solution in the long run (I was surprised that I could use preyEvolve, even though the evolved array was not instantiated and was causing a null pointer), so I left the beginnings of my other attempted solution commented out in case we need it later (I can delete it after talking to you about it tomorrow).

**Comment by Rollinsa** [comment]
Took a stab at implementing ObjectiveBestTeamsExperiment in order to have post best objective watch for cooperative coevolution. There was an error (which I commented out with a TODO) that I could not figure out, and a couple uncertainties (I used some code from the LonerTask post best objective watch experiment, ObjectiveBestNetworksExperiment, which I am not sure was necessary or if some needed changes or not) but I believe most of the needed implementation of it is there. Also note that, just like was done with ObjectiveBestNetworksExperiment with LonerTasks, I hard coded my ObjectiveBestTeamsExperiment to work for the CooperativeTask because it was necessary to get the number of populations, the objectives for each population, and also the evaluation of the best agent according to the corresponding task.

**Comment by Rollinsa** [comment]
Made the cooperative predPrey batch files have better names. Made a batch file for the postBestObjectiveWatch for cooperative predPrey and another one for the Eval for cooperative predPrey. Fixed a few bugs in the ObjectiveBestTeamsExperiment class, although there is now a new error on line 63 when trying to run postBestWatch, saying I can't cast the array to a Genotype[] array. Also, there is an error with the new post best eval batch file on line 48 of ObjectiveBestTeamsExperiment. Added a parameter for the  cooperative watch. Fixed several errors in the CooperativeTask, which now reveals another new error on line 190 of TorusPredPreyTask at the end of the first generation with monitorInputs and watch turned on. This error is because it uses the lonerTask "evolved" array which is null for cooperativeTask, but I am unsure as to how to avoid this error. Cleaned up CooperativeTorusPredPreyTask a bit.

**Comment by Rollinsa** [comment]
Fixed Batch files for cooperative watch and eval. Fixed and completed (hopefully) ObjectiveBestTeamsExperiment, so the postbest watch and eval should completely work for cooperative coevolution. Prevented childrenScore logging. Prevented extraneous panel with no sensor values, just the "individual info". Fixed error with monitorInputs by making evolved array global for cooperative getPredAgents and getPreyAgents. Made scoreSummary static so I could use it in the ObjectiveBestTeamsExperiment. Created an arrayList to array method for Genotypes in the PopulationUtil class.
